Dev

rinpmchs opened the pull request.

rinpmchs committed:
empty commit

rinpmchs committed:
create README.md

rinpmchs committed:
create .gitignore

rinpmchs committed:
add files

first
// i'm sorry for govnocode //

DimaTrushin commented on CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+project(LPN_algorithms)
+
+set(CMAKE_CXX_STANDARD 17)
+
```
Тут надо добавить
```cpp
set(CMAKE_CXX_STANDARD_REQUIRED True);
```
Иначе не гарантируется 17 стандарт. Cmake может откатиться до более младшего.

DimaTrushin commented on CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+set(CMAKE_CXX_STANDARD 17)
+
+set(Boost_INCLUDE_DIR /Users/ariko/boost_1_84_0)
+set(Boost_LIBRARY_DIR /Users/ariko/boost_1_84_0/libs)
```
Так делать плохо. У тебя машинозависимые пути. Любой другой пользователь не сможет запустить сборку твоего проекта, если не поменяет эти пути. Правильно пользоваться пакетным менеджером. У cmake тоже есть свои способы находить библиотеки, но find_package надо еще заморочиться правильно использовать.

DimaTrushin commented on main.cpp:
```diff
@@ -0,0 +1,9 @@
+int main()
+{
+//    Project::Impl::
+    Tests::run_all_tests();
```
Так как у тебя не отключены исключения, то по хорошему тут надо еще все обернуть в try/catch например так
```cpp
try {
  Tests::run__all_tests();
} catch(...) {
  Except::react();
}
```
где react это функция определенная в другом месте вида
```cpp
Except.h
namespace Except {
  void react();
}
Except.cpp
namespace Except {
void react() {
  try {
    throw;
  } catch(std::exception& e) {
    // Обработка известных исключений
  } catch(...) {
    // Обработка всех остальных исключений.
  }
}
}
```

DimaTrushin commented on project.h:
```diff
@@ -0,0 +1,27 @@
+#include <vector>
+#include <boost/multiprecision/cpp_int.hpp>
+
+using Type32_t = int32_t;
```
Не делай глобальных юзингов. Помести их все в Project namespace.

rinpmchs replied:
и везде, где я использую какой-то тип, писать Project::Type?

DimaTrushin commented on EratosthenesSieve/sieve.h:
```diff
@@ -0,0 +1,11 @@
+
+bool is_prime(const TypeU32_t& n, const std::vector<TypeU32_t>& v);
+
+}
```
Добавь пустую строку в конце каждого файла. Дело в том, что гит на это ругается и пользователь твоего репозитория будет видеть сообщения об ошибках от гита. Не делай так.

DimaTrushin commented on EratosthenesSieve/sieve.h:
```diff
@@ -0,0 +1,11 @@
+
+namespace Sieve {
+
+void sieve(const TypeU16_t& j, const TypeU32_t& n, std::vector<TypeU32_t>& a);
```
0) Плохие названия аргументов из имен j, n, a вообще не понятно что они значат и какой их смысл в алгоритме. Переименуй.
1) Встроенные типы не надо передавать по ссылке, передавай их только по значению. Я имею в виду TypeU16_t и TypeU32_t.
2) В том виде в котором у тебя введены псевдонимы не очень понятна их роль. Ты по сути завела другие имена для типов из stl, а в этом особенного смысла нет. Вот что такое у тебя j в качестве первого аргумента? Если это число, то было бы лучше увидеть NumberU16 или что-то в этом духе. Если это индекс, то Index или если хочешь подчеркнуть его детали, то IndexU32 или что-то в этом духе. Но по хорошему псевдоним для того и нужен, что ты не обращаешь внимания на детали.
3) Судя по всему у тебя j и n это входные параметры, а a выходное параметр в который ты пишешь. Входные надо передавать либо по константной ссылке либо по значению, а выходные по указателю и выходные всегда идут в конце. То есть тебе по хорошему надо a передавать по указателю. Это нужно для того, чтобы вызывающий код знал, какие из аргументов изменятся. Например сравни два вызова
```cpp
sieve(5, 6, a);
sieve(5, 6, &a);
```
Во втором случае сразу видно, что раз a передается по указателю, то она будет меняться.

DimaTrushin commented on EratosthenesSieve/sieve.cpp:
```diff
@@ -0,0 +1,49 @@
+
+namespace Sieve
+{
+void sieve(const TypeU16_t &j, const TypeU32_t &n, std::vector<TypeU32_t> &a)
```
Я вообще не понял, что эта функция делает.

rinpmchs replied:
они "просеивает" вектор и помечает все кратные данному простому числу числа как составные

DimaTrushin commented on EratosthenesSieve/sieve.cpp:
```diff
@@ -0,0 +1,49 @@
+
+std::vector<TypeU32_t> generate_primes(TypeU32_t n)
+{
+    std::vector<TypeU32_t> a(n + 1, 1);
```
Если ты ниже перезаписываешь данные в a, то сделай std::vector<TypeU32_t> a(n + 1); чтобы заполнить дефолтно.
Так же выбери нормальные имена для переменных, как я должен понять для чего эта переменная нужна по имени a?

rinpmchs replied:
здесь нужно все изначально пометить единицами, в новом коммите понятнее

DimaTrushin commented on EratosthenesSieve/sieve.cpp:
```diff
@@ -0,0 +1,49 @@
+{
+void sieve(const TypeU16_t &j, const TypeU32_t &n, std::vector<TypeU32_t> &a)
+{
+    TypeU16_t t = 2 * j;
```
А почему пропускается j?

rinpmchs replied:
потому что первое число, которое мы встречаем (после предыдущего шага), является простым, и нам нужно "вычеркнуть" только его кратные, а само это число оставить

DimaTrushin commented on EratosthenesSieve/sieve.cpp:
```diff
@@ -0,0 +1,49 @@
+    std::vector<TypeU32_t> a(n + 1, 1);
+    std::vector<TypeU32_t> v;
+
+    for (TypeU32_t i = 2; i <= n; ++i)
+    {
+        a[i] = i;
```
Ошибка уровня абстракции. По хорошему ты строишь вектор заполненный определенным образом. Эту работу ты тут не должна делать руками. Ты должна вызвать функцию, которая это сделает. Например так
```cpp
std::vector<TypeU32_t> a = ListOfNumbers(n + 1);
```

DimaTrushin commented on EratosthenesSieve/sieve.cpp:
```diff
@@ -0,0 +1,49 @@
+            sieve(j, n, a);
+        j++;
+    }
+
```
Я понял логику кода выше. Обычно делают не так. Обычно заводят массив
```cpp
std::vector<bool> is_prime(n);
```
И логика такая. По умолчанию is_prime[i] = false. И is_prime[i] соответствует i-ому нечетному числу начиная с 3. То есть если индекс i, то число для него будет p = 2 * i + 3. А если по числу надо посчитать индекс, то это будет i = (p - 3) / 2. И задача пройтись по этому массиву и установить флаг true, если определили, что число простое. Обычно в stl std::vector<bool> имеет особую сжатую имплементацию и у тебя будет уходить ровно 1 бит на один элемент вектора. Это позволяет сильно сэкономить по памяти.

Итоге у нас будет такая проверка, если is_prime[i] истино, то число 2 * i + 3 является простым. И уже эти числа можно засунуть в вектор.

DimaTrushin commented on EratosthenesSieve/sieve.h:
```diff
@@ -0,0 +1,11 @@
+
+std::vector<TypeU32_t> generate_primes(TypeU32_t n);
+
+bool is_prime(const TypeU32_t& n, const std::vector<TypeU32_t>& v);
```
Плохая сигнатура у метода. По хорошему ответ на вопрос простое ли число зависит только от числа, а не от числа и вектора чисел. По сути этот метод просто проверяет принадлежность контейнеру. И более того только вектору. Если бы был другой контейнер, то это бы не работало.

Как я вижу способ правильно доавить этот метод. Я бы создал класс
```cpp
class Tester {
public:
  bool is_prime(TypeU32_t number);
private:
  std::vector<TypeU32_t> primes_;
};
```
И метод проверки на простоту делает следующее. Он проверяет если sqrt{number} <= primes_back(), то мы просто проверяем, что number не делиться на все числа из primes_ меньше или равные sqrt(number). Если же эта оценка не выполняется, то сначала дополним список простых до sqrt(number) и потом запустим проверку.

DimaTrushin commented on FermatsAlgorithm/fermatsFact.h:
```diff
@@ -0,0 +1,39 @@
+
+}
+
+//
```
Не оставляй закомментированный код. У тебя специально есть система поддержки версий git, которая хранит историю всего, что ты делала. Если тебе надо поэксперементировать. Создаешь еще одну ветку
```
git checkout -b experiment1
```
и ты переходишь на новую ветку. В ней делаешь что хочешь, пишешь любой код, с которым хочешь поиграться, коммитишь его в эту ветку. Потом понимаешь, что надо вернуться обратно в master и пишешь
```
git checkout master
```
и продолжаешь работать в основной ветке. И с помощью checkout можно перенести из experiment1 ветки в твою любой нужный тебе файл командой
```
 git checkout experiment1 -- file.h
```

DimaTrushin commented on FermatsAlgorithm/fermatsFact.h:
```diff
@@ -0,0 +1,39 @@
+
+namespace FermatFact {
+
+TypeU64_t squareroot(const TypeU64_t& n);
```
По хорошему эта функция это просто вспомогательная математическая функция. Она не относится к алгоритму факторизации. Ты ее можешь захотеть использовать еще где-то. Потому ее стоит выделить отдельно.

DimaTrushin commented on FermatsAlgorithm/fermatsFact.cpp:
```diff
@@ -0,0 +1,39 @@
+
+TypeU64_t squareroot(const TypeU64_t& n) {
+    TypeU64_t sqrt_n = (TypeU64_t) std::sqrt(n);
+    if (sqrt_n ==  std::sqrt(n))
```
Я всегда с опаской смотрю на такие условия. double и сравнение на равенство -- большие враги. Не факт, что это сравнение не выдаст тебе ошибочный результат. Надо смотреть как имплементирована sqrt или какие она дает гарантии. Надежнее всего всегда прибавить дополнительно.

DimaTrushin commented on FermatsAlgorithm/fermatsFact.cpp:
```diff
@@ -0,0 +1,39 @@
+    v = 1;
+    Type64_t r = sqrt_n * sqrt_n - n;
+
+    // X_LOOP
```
- Со следующим кодом есть несколько проблем. У тебя два вложенных цикла да еще и по одной и той же переменной r. Это означает, что у тебя на все эти циклы общий инвариант. И хрен знает какой ой.
- Кроме того, у тебя тут 4 вложенных scope если считать вместе с функцией. Даже в ядре Linux разрешено не более 3.
- У тебя перед каждым циклом или if-ом идет комментарий. Это значит что ты хочешь словами выделить функцию, которую не выделяешь кодом.

В сухом остатке. Надо переписать более нормально эту функцию.

DimaTrushin commented on FermatsAlgorithm/fermatsFact.cpp:
```diff
@@ -0,0 +1,39 @@
+            u += 2;
+        }
+    }
+//    assert ((u - v) / 2 == 1)
```
Почему assert закомментирован?

DimaTrushin commented on project.h:
```diff
@@ -0,0 +1,27 @@
+using TypeU32_t = uint32_t;
+using TypeU64_t = uint64_t;
+using TypeU128_t = boost::multiprecision::uint128_t;
+using TypeInteger = boost::multiprecision::cpp_int;
```
Лучше назови LongInteger, чтобы было понятно, что это длинка. Слово Type в названии псевдонима не стоит указывать. Мы и так понимаем, что это тип, а компилятору от этого нет проку, он не поможет если мы ошиблись.

DimaTrushin commented on Useful/algorithms.h:
```diff
@@ -0,0 +1,13 @@
+
+namespace UsefulAlgorithms {
+
+TypeU16_t power_mod(TypeU16_t a, TypeU16_t b, TypeU16_t m);
+
+TypeU32_t power_mod(TypeU32_t a, TypeU32_t b, TypeU32_t n);
+
+TypeU64_t power_mod(TypeU64_t a, TypeU64_t b, TypeU64_t m);
+
+TypeInteger power_mod(TypeInteger a, TypeInteger b, TypeInteger m);
```
Типы данных у длинки могту быть тяжелые. Потому плохая идея принимать их by value. Лучше by const reference: const LongInteger&.

DimaTrushin replied:
У тебя все эти функции одинаково имплементируются. Тебе можно сделать шаблонную функцию, которая делает эту работу за всех, чтобы избежать дублирования кода.

DimaTrushin commented on PollardsAlgorithm/pollardAlg.cpp:
```diff
@@ -0,0 +1,24 @@
+        m = UsefulAlgorithms::power_mod(m, i , n);
+        if (i % 10 == 0) {
+            // CHECK_GCD
+            TypeInteger g = boost::math::gcd(m - 1, n);
```
Плохой дизайн. Ты с одной стороны ввела свой псевдоним для типа, а потом для него явно зовешь бустовсткую функцию. Если ты замениш TypeInteger на что-то другое, тогда тебе и тут придется поменять вызов на другую gcd. Лучше в useful добавь gcd для TypeInteger и внутри него вызови boost gcd.

DimaTrushin commented on Pseudoprimes/pseudoprimes.h:
```diff
@@ -0,0 +1,7 @@
+
+namespace PseudoPrimes {
+
+bool is_pseudoprime();
```
Офигенный метод. Он ничего не принимает, но возвращает значения. Это сразу говорит, что у тебя есть скрытые глобальные данные, а это плохо.

DimaTrushin commented on Pseudoprimes/pseudoprimes.cpp:
```diff
@@ -0,0 +1,18 @@
+namespace PseudoPrimes {
+
+bool is_pseudoprime() {
+    TypeU32_t n = 341; // n is odd and composite
```
Почему это число не принимается аргументом?

rinpmchs replied:
это просто полуначатая функция, извини за мусор

DimaTrushin commented on TrialDivision/trialDiv.h:
```diff
@@ -0,0 +1,12 @@
+
+namespace TrialDivision {
+
+int divide(TypeU64_t& f, TypeU32_t d, int& i);
```
а зачем нужно i?

rinpmchs replied:
убрала

DimaTrushin commented on project.h:
```diff
@@ -0,0 +1,27 @@
+#include <vector>
+#include <boost/multiprecision/cpp_int.hpp>
+
+using Type32_t = int32_t;
```
У тебя страшная каша из типов. Сделай для чисел без длинки пару юзингов
```cpp
using SmallInteger = int64_t;
using SmallUInteger = uint64_t;
```
И вообще не используй больше ничего для чисел. Потому что у тебя каша из кода с разными типами целых и не понятно что где и когда зовется и все ли случаи покрыты и так далее.

DimaTrushin commented on TrialDivision/trialDiv.cpp:
```diff
@@ -0,0 +1,54 @@
+    TypeU64_t f = n; // still unfactored portion
+    std::map<TypeU64_t , int> factors;
+
+    // TRY_2&3
```
Куда читаемее было бы без цикла. И можно не писать комментарий. Я вижу, что здесь именно это и происходит.

DimaTrushin commented on TrialDivision/trialDiv.cpp:
```diff
@@ -0,0 +1,54 @@
+
+// works fine for numbers up to ten or eleven digits --> TypeU32_t is not enough
+std::map<TypeU64_t , int> factorize(const TypeU64_t& n, const TypeU64_t& max) {
+    int i = 0; // number of distinct prime factors
```
Зачем этот счетчик, если у тебя есть factors и он имеетметод size()?

rinpmchs replied:
согласна! списала это втупую из книжки, не заметила, что он бессмысленный

DimaTrushin commented on TrialDivision/trialDiv.cpp:
```diff
@@ -0,0 +1,54 @@
+    return factors;
+}
+
+bool is_composite(TypeU64_t n) {
```
У тебя этот метод повторяется в разных местах по нескольку раз но с разными названиями.

DimaTrushin commented on Tests/testAlgorithms.h:
Во всех файлах с тестами тебе не надо включать ничего в хедеры. Например include здесь в первой строчке можно перенести в testAlgorithms.cpp. И будет только лучше инкапсюляция. main не придется перекомпилировать вообще никогда.

DimaTrushin reviewed (changes requested):
Обрати внимание, что я все комментарии делал разово. Если какая-то проблема встречалась несколько раз тебе надо самой исправлять по аналогии.

В целом все нормально, но есть проблемы.

1) У тебя очень много разных целых типов с юзингами и ты непойми для каких типов что-то делаешь. Я бы использовал всего 2 (или 4 если учитывать беззнаковые) юзинга для чисел без длинки и чисел с длинкой и потом все писал бы только для них. Более того, часто у тебя код одинаковый что для длинки, что без нее. Потому можно все переписать на шаблонах, но не обязательно.

2) Вынеси все вспомогательные функции, которые не привязаны к методам и алгоритмам отдельно. У тебя некоторые методы по десять раз повторяются (условно).

3) Ты группируешь все по namespace-ам. А есть другой впособ группировать методы внутри классов и структур. Тогда если им приходится разделять какие-то данные, то их можно не делать глобальными. Это решит проблему с передачей глобальных данных.

4) Еще местами у тебя странный код с кучей комментариев. Комментариев быть не должно. Единственное, я позволю тебе иметь комментарии в run_all_tests(). Выделяй функции с нужными именами вместо комментариев.

rinpmchs committed:
add new algorithms

checkpoint 2: AKS and Dijkstra added
[AKS doesn't work with primes >= 67; Dijkstra is not finished]

rinpmchs committed:
update sieve

rinpmchs committed:
make sieve readable

rinpmchs committed:
add "tests"

These "tests" are used to test if functions work by the user. 
Later i will rename this folder to UseCases or UsageExamples^ and make Tests using gtests

rinpmchs committed:
make FermatsFact class

rinpmchs committed:
add use cases

they were "tests" before

rinpmchs committed:
add google tests + data

rinpmchs committed:
delete "tests"

they are now use cases

rinpmchs committed:
add tests and data

rinpmchs committed:
delete trash

rinpmchs committed:
update CMakeLists.txt

rinpmchs committed:
add tests

write rusults to file option added

rinpmchs committed:
add more tests + results

rinpmchs committed:
add combinations

rinpmchs committed:
delete empty files

rinpmchs committed:
add use combinations

(example of using combinations with prime and composite numbers)

rinpmchs committed:
update main.cpp

rinpmchs committed:
update project

rinpmchs committed:
update tests and results

rinpmchs committed:
update useful

rinpmchs committed:
add integer.h

rinpmchs committed:
delete project.h

rinpmchs committed:
delete project.cpp

rinpmchs committed:
update main.cpp

rinpmchs committed:
delete AKSAlgorithm

rinpmchs committed:
update combinations

rinpmchs committed:
update sieve

rinpmchs committed:
update exceptions

rinpmchs committed:
update fermat

rinpmchs committed:
add miller-rabin

rinpmchs committed:
update pollard

rinpmchs committed:
delete trash

rinpmchs committed:
update tests

rinpmchs committed:
update trial division

rinpmchs committed:
update use cases

rinpmchs committed:
update CMakeLists.txt

rinpmchs committed:
add conanfile.txt

DimaTrushin commented on Useful/profiler.cpp:
```diff
@@ -0,0 +1,28 @@
+std::chrono::duration<double, std::milli> Profiler::totalExecTimeDbl;
+
+void Profiler::start() {
+    Profiler::executionStartTimestamp = std::chrono::high_resolution_clock::now();
```
Вот тут не надо Profiler::, эта переменная и так будет видна.

DimaTrushin commented on Useful/profiler.h:
```diff
@@ -0,0 +1,19 @@
+
+namespace Project::Useful {
+class Profiler {
+public:
```
можно внутри профайлера сделать 
```cpp
namespace chrono = std::chrono;
```
И после этого можно будет писать просто chrono::something вместо std::chrono::something.

DimaTrushin commented on Useful/profiler.h:
```diff
@@ -0,0 +1,19 @@
+    static double getExecutionTimeDouble();
+
+private:
+    static std::chrono::time_point<std::chrono::high_resolution_clock> executionStartTimestamp;
```
Не используй такие сложные имена в типах. Заводи для них using-ги для повышения читаемости.

DimaTrushin commented on Useful/algorithms.h:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    LongInteger generate_number() {
+        return this->ui(this->mt);
```
Не используй явно this в коде. Тут можно написать
```cpp
return ui_(mt_);
```

DimaTrushin commented on Useful/algorithms.h:
```diff
@@ -0,0 +1,30 @@
+
+private:
+    boost::random::mt19937 mt;
+    boost::random::uniform_int_distribution<LongInteger> ui;
```
Добавь _ в конце имен переменных.

DimaTrushin commented on Useful/algorithms.h:
```diff
@@ -0,0 +1,30 @@
+        this->ui = boost::random::uniform_int_distribution<LongInteger>(from_margin, to_margin);
+    }
+
+    LongInteger generate_number() {
```
Почему имплементация в хедере а не в cpp? перенести

rinpmchs replied:
конструктор тоже в cpp перенести?

DimaTrushin commented on Useful/algorithms.h:
```diff
@@ -0,0 +1,30 @@
+public:
+    RandomGenerator(const LongInteger& from_margin, const LongInteger& to_margin) {
+        // set range for uniform int distribution - ui
+        this->ui = boost::random::uniform_int_distribution<LongInteger>(from_margin, to_margin);
```
Вся инициализация должна быть в списке инициализации, то есть надо
```cpp
RandomGenerator(const LongInteger& from, const LongInteger& to)
  : mt_(42), ui_(boost::random::uniform_int_distribution<LongInteger>(from, to)) {
}
```

DimaTrushin commented on Useful/algorithms.cpp:
```diff
@@ -0,0 +1,80 @@
+            return true;
+        d++;
+    }
+    return false; // n is prime or pseudoprime
```
Если у тебя ответ бывает не просто да/нет, то лучше не возвращать bool. Заведи enum class Status {Prime, Composite, Pseudoprime}; и возвращай статус операции.

rinpmchs replied:
тут непонятно, prime или pseudoprime
этот комментарий значит, что тест неидеален и есть вероятность ошибки

DimaTrushin commented on Useful/algorithms.cpp:
```diff
@@ -0,0 +1,80 @@
+    return false; // n is prime or pseudoprime
+}
+
+bool is_composite(SmallUInteger n) {
```
У тебя все методы имеют одинаковую имплементацию. Ты можешь заменить три метода на 1 шаблонный метод и не надо будет писать один и тот же код три раза.

DimaTrushin commented on UseCases/useTrialDivision.cpp:
```diff
@@ -0,0 +1,32 @@
+        std::cout << "n is prime or pseudoprime, n = ";
+    else
+        std::cout << "n is composite, n = ";
+    for (std::map<LongInteger, int>::iterator it = factors.begin(); it != --factors.end(); it++)
```
Чтобы не писать такие длинные имена итераторов можно поступать несколькими способами.
1) Завести для контейнера более короткий псевдоним и потом для итератора
2) Использовать ключевое слово auto.
3) Использовать range for loop.

DimaTrushin commented on UseCases/useTrialDivision.cpp:
```diff
@@ -0,0 +1,32 @@
+    LongInteger n, max;
+    std::cout << "input number and max divisor: ";
+    std::cin >> n >> max;
+    Detail::TrialDivision tr_div;
```
Старайся не использовать такие доступы через квалификатор имен для своих типов. Заводи using-ги перед использованием.

DimaTrushin commented on UseCases/useSieve.cpp:
```diff
@@ -0,0 +1,27 @@
+    std::cout << "number to check: ";
+    std::cin >> n;
+    Detail::Sieve sieve;
+//    std::vector<SmallInteger> v = std::move(generate_primes(n));
```
Убери весь закомментированный код, который не нужен для работы программы.

DimaTrushin commented on UseCases/useSieve.cpp:
```diff
@@ -0,0 +1,27 @@
+#include "useSieve.h"
+
+namespace Project::Use {
+void use_sieve_print_primes() {
```
1) Очень длинные имена методов. Слово use в начале лишнее, потому что ты и так будешь делать доступ через имя namespace-а, а оно как раз Use, то есть у тебя будет Use::use_sieve_print_primes. Потому use можно выкинуть.
2) Название функции должно начинаться с глагола и содержать только один глагол (не больше и не меньше). Иначе НЕ понятно, что делает твой метод.

Я думаю что можно метод назвать print_primes. Если у тебя есть другие методы print_primes. То можешь завернуть тут в Sieve namespace и будешь обращаться к методу Use::Sieve::print_primes и будет лучше.

DimaTrushin commented on UseCases/usePollard.h:
```diff
@@ -0,0 +1,3 @@
+namespace Project::Use {
+void use_pollard_factorization();
```
Плохое название метода. НЕ ясно что он делает.

DimaTrushin commented on UseCases/useSieve.h:
```diff
@@ -0,0 +1,4 @@
+namespace Project::Use {
```
Добавь в начале КАЖДОГО хедера #pragma once

DimaTrushin commented on UseCases/useFermat.cpp:
```diff
@@ -0,0 +1,23 @@
+    std::cout << "number to check: ";
+    std::cin >> n;
+
+    Useful::Profiler::start();
```
Заведи в начале функции using
```cpp
using Profiler = Useful::Profiler;
```
И потом сможешь пользваться Profiler::start(); и не писать длинных имен.

DimaTrushin commented on UseCases/usages.cpp:
```diff
@@ -0,0 +1,32 @@
+
+    try {
+        use_combination1();
+//        use_combination2();
```
Зачем тут закомментированный код?

DimaTrushin commented on UseCases/testAKS.cpp:
```diff
@@ -0,0 +1,23 @@
+}
+
+}  // namespace Project
+// что-то не работает, начиная с 67, хз почему
```
Это что за чудесный комментарий?

DimaTrushin replied:
Добавь в конце каждого файла пустую строку.

DimaTrushin commented on UseCases/testAKS.cpp:
```diff
@@ -0,0 +1,23 @@
+
+    std::cout << "number to check: ";
+    std::cin >> n;
+//    assert (typeid(n).name() == Project::SmallUInteger);
```
Почему закомментировано?

DimaTrushin commented on TrialDivision/trialDiv.cpp:
```diff
@@ -0,0 +1,50 @@
+
+bool TrialDivision::is_composite(SmallUInteger n) {
+    int i = 2;
+    while (i * i <= n) {
```
здесь можно вычислить корень из n потом преобразовать его в SmallUInteger и прибавить 1. И будет быстрее сравнивать i < sqrt_n

DimaTrushin commented on Tests/write_test_results.h:
```diff
@@ -0,0 +1,23 @@
+
+namespace Tests {
+
+extern std::ofstream test_results;
```
Это пиздец. Нахуя, а главное зачем? Убрать глобальные переменные. Сделай это переменной в той функции, которой ты пользуешься. Или сделай это переменной класса, на худой конец статической переменной класса, но НЕ extern глобальной переменной!

DimaTrushin commented on Tests/tests.cpp:
```diff
@@ -0,0 +1,16 @@
+    } catch (...) {
+        Except::handleException("gtests run failed");
+    }
+    return -1;
```
Не возвращай -1. Возвращай 0.

DimaTrushin commented on FermatsAlgorithm/fermatsFact.h:
```diff
@@ -0,0 +1,8 @@
+namespace Project::Detail {
+class FermatFact {
+public:
+    std::pair<SmallUInteger, SmallUInteger> factorize(const SmallUInteger& n);
```
Не используй std::pair. Заведи структуру с нормальным именем. Вот что значит first и second? Если это делитель и его степень, то сделай
```cpp
templte<class Int>
struct FactorData {
  Int factor;
  int power;
};
```

DimaTrushin commented on EratosthenesSieve/sieve.h:
```diff
@@ -0,0 +1,11 @@
+class Sieve {
+public:
+    std::vector<SmallInteger> generate_primes(SmallInteger n);
+    bool check_primality(SmallInteger n, const std::vector<SmallInteger>& primes);
```
Почему метод не константный?

DimaTrushin commented on EratosthenesSieve/sieve.h:
```diff
@@ -0,0 +1,11 @@
+namespace Project::Detail {
+class Sieve {
+public:
+    std::vector<SmallInteger> generate_primes(SmallInteger n);
```
Почему метод НЕ константный?

DimaTrushin commented on EratosthenesSieve/sieve.h:
```diff
@@ -0,0 +1,11 @@
+    std::vector<SmallInteger> generate_primes(SmallInteger n);
+    bool check_primality(SmallInteger n, const std::vector<SmallInteger>& primes);
+private:
+    void sieve(SmallInteger p, SmallInteger n, std::vector<bool>& is_prime);
```
Почему метод НЕ константный?

DimaTrushin commented on Combinations/combination2.h:
```diff
@@ -0,0 +1,11 @@
+{
+public:
+    friend class Comb1;
+    bool comb2_primality_test(const LongInteger &number);
```
Убери название comb2 из этой функции. Это уже есть в названии класса. Начни функцию с глагола.

DimaTrushin commented on Combinations/combination2.cpp:
```diff
@@ -0,0 +1,31 @@
+        return false;
+
+    Sieve sieve;
+    SmallInteger limit = 1000;
```
Все константы сделай статическими constexpr выражениями приватными в классе.

DimaTrushin commented on Combinations/combination1.h:
```diff
@@ -0,0 +1,15 @@
+class Comb1
+{
+public:
+    friend class MillerRabin;
+    friend class Comb2;
```
УБРАТЬ! Никаких friend-ов быть НЕ должно. Сделай нормальные интерфейсы для взаимодействия.

DimaTrushin commented on integer.h:
```diff
@@ -0,0 +1,14 @@
+using SmallUInteger = uint64_t;
+using LongInteger = boost::multiprecision::cpp_int;
+
+namespace Detail {}
+namespace Tests {}
+namespace Use {}
+namespace Useful {}
```
Убрать. Они тут НЕ нужны. Пространства имен можно открывать где угодно.

DimaTrushin reviewed (changes requested):
Обрати внимание, что я не повторял замечания. Если одна и та же ошибка всречалась несколько раз, я комментировал лишь единожды. Потому я ожидаю, что ты поправишь все аналогичные замечания всюду.

rinpmchs committed:
delete useful

rinpmchs committed:
add utils

rinpmchs committed:
update use cases

rinpmchs committed:
update trial division

rinpmchs committed:
update tests

rinpmchs committed:
update pollard

rinpmchs committed:
update miller rabin

rinpmchs committed:
update fermat

rinpmchs committed:
update exceptions

rinpmchs committed:
update sieve

rinpmchs committed:
update combinations

rinpmchs committed:
update main.cpp

rinpmchs committed:
update integer.h

rinpmchs committed:
update CMakeLists.txt

rinpmchs committed:
update trial division

rinpmchs committed:
update fermat
